traversal/amend/amender.go

```diff
@@ -0,0 +1,147 @@
+package amend
+
+import (
+	"fmt"
+	"github.com/ipld/go-ipld-prime/datamodel"
+	"github.com/ipld/go-ipld-prime/node/basicnode"
+)
+
+type Amd string
+
+const (
+	Amd_Add    = "add"
+	Amd_Remove = "remove"
+)
+
+type Amendment struct {
+	Op    Amd
+	Path  datamodel.Path
+	Value datamodel.Node
+}
+
+// -- Node -->
+
+var _ datamodel.Node = (AmenderNode)(nil)
+
+type AmenderNode = *amenderNode
+
+type amenderNode struct {
+	base   datamodel.Node
+	addOps []Amendment
+	remOps map[string]Amendment
+}
+
+func NewAmender(base datamodel.Node) AmenderNode {
+	return &amenderNode{base: base}
+}
+
+func (*amenderNode) Kind() datamodel.Kind {
+	return datamodel.Kind_Map
+}
+func (*amenderNode) LookupByString(key string) (datamodel.Node, error) {
+	panic("misuse")
+}
+func (*amenderNode) LookupByNode(datamodel.Node) (datamodel.Node, error) {
+	panic("misuse")
+}
+func (*amenderNode) LookupByIndex(idx int64) (datamodel.Node, error) {
+	panic("misuse")
+}
+func (*amenderNode) LookupBySegment(seg datamodel.PathSegment) (datamodel.Node, error) {
+	panic("misuse")
+}
+func (a *amenderNode) MapIterator() datamodel.MapIterator {
+	if a.base.Kind() != datamodel.Kind_Map {
+		panic("misuse")
+	}
+	return &amender_Iterator{a, a.base.MapIterator(), 0}
+}
+func (*amenderNode) ListIterator() datamodel.ListIterator {
+	panic("misuse")
+}
+func (a *amenderNode) Length() int64 {
+	return a.base.Length() + int64(len(a.addOps)) - int64(len(a.remOps))
```

Comment:
I don't think this can be accurate as far as a plain node is concerned, since we don't know anything about overlap and conflicts between add and remove ops (e.g. add a field twice); but maybe it doesn't matter here. But what would be the purpose or reporting a `Length()` that has this base+add-remove calculation?

Reply:
Coincidentally, I was just refactoring this :) I'm thinking of keeping a single map of mods. That way, the "effective" length calculation as well as the conflict resolution between adds/removes are simpler.

Reply:
A non-`nil` value in the mods map means that a node was added (or added/removed/re-added, and so on), while a value of `nil` means that it was removed.

Reply:
The currently pushed code definitely has bugs that I'm fixing (please don't hesitate to point them out). Correcting it to be more intelligent w.r.t. types of nodes.

Reply:
@rvagg, I've iterated a couple of times and pushed another update with all (current) fixtures passing for map-type nodes.

It has a similar base+add-remove type calculation as before (with some adjustments) so that `Encode` sees an accurate "effective length" for recursive nodes.